runtime/signals_nat.c

```diff
@@ -57,8 +59,52 @@ extern signal_handler caml_win32_signal(int sig, signal_handler action);
 
 void caml_garbage_collection()
 {
+  frame_descr* d;
+  intnat allocsz = 0;
+  char *sp;
+  uintnat retaddr;
+  intnat whsize;
+
+  caml_frame_descrs fds = caml_get_frame_descrs();
+  struct stack_info* stack = Caml_state->current_stack;
+
+  sp = (char*)stack->sp;
+  retaddr = *(uintnat*)sp;
+
+  { /* Find the frame descriptor for the current allocation */
+    uintnat h = Hash_retaddr(retaddr, fds.mask);
+    while (1) {
+      d = fds.descriptors[h];
+      if (d->retaddr == retaddr) break;
+      h = (h + 1) & fds.mask;
+    }
+    /* Must be an allocation frame */
+    CAMLassert(d && d->frame_size != 0xFFFF && (d->frame_size & 2));
+  }
+
+  { /* Compute the total allocation size at this point,
+       including allocations combined by Comballoc */
+    unsigned char* alloc_len = (unsigned char*)(&d->live_ofs[d->num_live]);
+    int i, nallocs = *alloc_len++;
+    for (i = 0; i < nallocs; i++) {
+      /* Since 2 words is the smallest allocation, sizes are
+         encoded as (wosize - 2).
+         See Emitaux.emit_frames and caml/stack.h */
+      allocsz += alloc_len[i] + 2;
+    }
+    /* We have computed whsize (including header), but need wosize (without) */
+    allocsz -= 1;
+  }
+
+  whsize = Whsize_wosize(allocsz);
+
+  Caml_state->young_ptr += whsize * sizeof(value);
+
   caml_handle_gc_interrupt();
 
+  /* Re-do the allocation: we now have enough space in the minor heap. */
+  Caml_state->young_ptr -= whsize * sizeof(value);
```

Comment:
Are we able to assert that their really is enough space in the minor heap?

The case I'm worried about is if somehow we end up in a position after `caml_handle_gc_interrupt` where:
- a minor collection happens
- major slice work schedules a finaliser
- we run the finaliser which fills up the minor heap so there isn't enough space but doesn't trigger a collection